### PR TITLE
ssz: fix ssv sanity tests

### PR DESCRIPTION
**Pull request by nkryuchkov**

### CODE REVIEW

**Review by aaron-blox — APPROVED** [review]

File: qbft/spectest/tests/messages_spectest.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ func (test *MsgSpecTest) Run(t *testing.T) {
 			continue
 		}
 
-		if msg.Message.MsgType == qbft.RoundChangeMsgType && len(msg.Message.RoundChangeJustification) == 0 {
+		if msg.Message.RoundChangePrepared() && len(msg.Message.RoundChangeJustification) == 0 {
```
Comment (alonmuroch) [inline qbft/spectest/tests/messages_spectest.go:28]: what did you try to do here?

File: qbft/spectest/tests/proposal/invalid_prepare_justification_round.go [hunk 2]
```diff
@@ -18,9 +18,12 @@ func InvalidPrepareJustificationRound() *tests.MsgProcessingSpecTest {
 		testingutils.TestingPrepareMessageWithRound(ks.Shares[3], types.OperatorID(3), 2),
 	}
 	rcMsgs := []*qbft.SignedMessage{
-		testingutils.TestingRoundChangeMessageWithRound(ks.Shares[1], types.OperatorID(1), 2),
-		testingutils.TestingRoundChangeMessageWithRound(ks.Shares[2], types.OperatorID(2), 2),
-		testingutils.TestingRoundChangeMessageWithRound(ks.Shares[3], types.OperatorID(3), 2),
+		testingutils.TestingRoundChangeMessageWithRoundAndRC(ks.Shares[1], types.OperatorID(1), 2,
```
Comment (alonmuroch) [inline qbft/spectest/tests/proposal/invalid_prepare_justification_round.go:21]: why this change?

File: qbft/spectest/tests/proposal/prepared_previously_no_prepare_justification_quorum.go [hunk 3]
```diff
@@ -17,9 +17,12 @@ func PreparedPreviouslyNoPrepareJustificationQuorum() *tests.MsgProcessingSpecTe
 		testingutils.TestingPrepareMessage(ks.Shares[2], types.OperatorID(2)),
 	}
 	rcMsgs := []*qbft.SignedMessage{
-		testingutils.TestingRoundChangeMessageWithRound(ks.Shares[1], types.OperatorID(1), 2),
-		testingutils.TestingRoundChangeMessageWithRound(ks.Shares[2], types.OperatorID(2), 2),
-		testingutils.TestingRoundChangeMessageWithRound(ks.Shares[3], types.OperatorID(3), 2),
+		testingutils.TestingRoundChangeMessageWithRoundAndRC(ks.Shares[1], types.OperatorID(1), 2,
```
Comment (alonmuroch) [inline qbft/spectest/tests/proposal/prepared_previously_no_prepare_justification_quorum.go:20]: why this change?

File: qbft/spectest/tests/roundchange/duplicate_msg_quorum_prepared_rc_first.go [hunk 4]
```diff
@@ -19,7 +19,8 @@ func DuplicateMsgQuorumPreparedRCFirst() *tests.MsgProcessingSpecTest {
 		testingutils.TestingPrepareMessage(ks.Shares[3], types.OperatorID(3)),
 	}
 	msgs := []*qbft.SignedMessage{
-		testingutils.TestingRoundChangeMessageWithRound(ks.Shares[1], types.OperatorID(1), 2),
+		testingutils.TestingRoundChangeMessageWithRoundAndRC(ks.Shares[1], types.OperatorID(1), 2,
```
Comment (alonmuroch) [inline qbft/spectest/tests/roundchange/duplicate_msg_quorum_prepared_rc_first.go:22]: why change?

File: qbft/spectest/tests/roundchange/f1_different_future_rounds.go [hunk 5]
```diff
@@ -27,10 +27,11 @@ func F1DifferentFutureRounds() *tests.MsgProcessingSpecTest {
 	return &tests.MsgProcessingSpecTest{
 		Name:          "round change f+1 prepared",
 		Pre:           pre,
-		PostRoot:      "342d0646bc8da4112539b97c68f21ef98bb784a2f97923bc1a81ff92656b8e92",
+		PostRoot:      "141d47f572e1620701feab85a1131206783f540e418693fabb80992a68ebaf16",
 		InputMessages: msgs,
 		OutputMessages: []*qbft.SignedMessage{
-			testingutils.TestingRoundChangeMessageWithRound(ks.Shares[1], types.OperatorID(1), 5),
+			testingutils.TestingRoundChangeMessageWithParams(ks.Shares[1], types.OperatorID(1), 5, qbft.FirstHeight,
```
Comment (alonmuroch) [inline qbft/spectest/tests/roundchange/f1_different_future_rounds.go:33]: why change?

File: qbft/spectest/tests/roundchange/f1_different_future_rounds_not_prepared.go [hunk 6]
```diff
@@ -20,10 +20,11 @@ func F1DifferentFutureRoundsNotPrepared() *tests.MsgProcessingSpecTest {
 	return &tests.MsgProcessingSpecTest{
 		Name:          "round change f+1 not prepared",
 		Pre:           pre,
-		PostRoot:      "6fff8e31d81c7f0121d00a1b6d16c66f0e733eed38ade739cad139d8fd63592f",
+		PostRoot:      "d8a2e893fc55a2aa3232d2c36a578260efa73cd083eecea1babec33c23725ba4",
 		InputMessages: msgs,
 		OutputMessages: []*qbft.SignedMessage{
-			testingutils.TestingRoundChangeMessageWithRound(ks.Shares[1], types.OperatorID(1), 5),
+			testingutils.TestingRoundChangeMessageWithParams(ks.Shares[1], types.OperatorID(1), 5, qbft.FirstHeight,
```
Comment (alonmuroch) [inline qbft/spectest/tests/roundchange/f1_different_future_rounds_not_prepared.go:26]: why change?

File: qbft/spectest/tests/roundchange/justification_wrong_value.go [hunk 7]
```diff
@@ -1,34 +0,0 @@
-package roundchange
-
-import (
```
Comment (alonmuroch) [inline qbft/spectest/tests/roundchange/justification_wrong_value.go:3]: why deleted test?

File: qbft/spectest/tests/roundchange/quorum_msg_not_prepared.go [hunk 8]
```diff
@@ -28,11 +28,12 @@ func QuorumMsgNotPrepared() *tests.MsgProcessingSpecTest {
 	return &tests.MsgProcessingSpecTest{
 		Name:          "round change quorum msg not prepared",
 		Pre:           pre,
-		PostRoot:      "7a694445a4a45867d4b1158b0cdaa92e5e1b6e5541a772ee7f30ce7e85947b93",
+		PostRoot:      "ce52b74e850002c75ea9a52749285b8778c0411989e2e0547547ed26bdb31211",
 		InputMessages: msgs,
 		OutputMessages: []*qbft.SignedMessage{
-			testingutils.TestingProposalMessageWithRoundAndRC(ks.Shares[1], types.OperatorID(1), 2,
-				testingutils.MarshalJustifications(prepareMsgs)),
+			testingutils.TestingProposalMessageWithParams(ks.Shares[1], types.OperatorID(1), 2, qbft.FirstHeight,
```
Comment (alonmuroch) [inline qbft/spectest/tests/roundchange/quorum_msg_not_prepared.go:34]: why change?

File: qbft/spectest/tests/timeout/test.go [hunk 9]
```diff
@@ -43,10 +45,14 @@ func (test *SpecTest) Run(t *testing.T) {
 		require.Len(t, broadcastedMsgs, len(test.OutputMessages))
 
 		for i, msg := range test.OutputMessages {
+			bytes1, _ := json.Marshal(msg)
+			log.Println("want: ", string(bytes1))
```
Comment (alonmuroch) [inline qbft/spectest/tests/timeout/test.go:49]: why we need printing?

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: qbft/spectest/tests/proposal/invalid_prepare_justification_round.go (hunk 2 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/proposal/invalid_prepare_justification_round.go:21]: This message has a justification on `main` branch (https://github.com/bloxapp/ssv-spec/blob/main/qbft/spectest/tests/proposal/invalid_prepare_justification_round.go#L43), but in a previous PR I made a mistake and removed it in some of the tests

File: qbft/spectest/tests/timeout/test.go (hunk 9 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/timeout/test.go:49]: Committed by mistake, sorry

File: qbft/spectest/tests/proposal/prepared_previously_no_prepare_justification_quorum.go (hunk 3 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/proposal/prepared_previously_no_prepare_justification_quorum.go:20]: due to my mistake in a previous PR (https://github.com/bloxapp/ssv-spec/pull/162/files#r1120058796)

File: qbft/spectest/tests/roundchange/duplicate_msg_quorum_prepared_rc_first.go (hunk 4 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/roundchange/duplicate_msg_quorum_prepared_rc_first.go:22]: mistake in a previous PR (https://github.com/bloxapp/ssv-spec/pull/162/files#r1120058796)

File: qbft/spectest/tests/roundchange/quorum_msg_not_prepared.go (hunk 8 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/roundchange/quorum_msg_not_prepared.go:34]: a mistake in a previous PR, this message should have both justifications according to https://github.com/bloxapp/ssv-spec/blob/main/qbft/spectest/tests/roundchange/quorum_msg_not_prepared.go#L73

File: qbft/spectest/tests/roundchange/justification_wrong_value.go (hunk 7 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/roundchange/justification_wrong_value.go:3]: I thought it needs to be deleted because we don't have the justification data field. Should I change the test logic instead of deleting it so that the test checks for invalid full data?

File: qbft/spectest/tests/messages_spectest.go (hunk 1 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/messages_spectest.go:28]: It's for `messages.RoundChangeDataInvalidJustifications` (https://github.com/bloxapp/ssv-spec/blob/ssz-fix-remaining-tests-issues/qbft/spectest/tests/messages/rc_invalid_justifications.go#L11). 

We used to have this check in https://github.com/bloxapp/ssv-spec/blob/main/qbft/messages.go#L140 but now we don't have it when validating a message (https://github.com/bloxapp/ssv-spec/blob/ssz-fix-remaining-tests-issues/qbft/messages.go#L122). 

I think I need to remove this check because it's in a test runner and doesn't actually test what we need to test.

Should we add a justification length check of prepared messages in https://github.com/bloxapp/ssv-spec/blob/ssz-fix-remaining-tests-issues/qbft/messages.go#L122 or should we just remove this check and remove the `messages.RoundChangeDataInvalidJustifications` test?

File: ssv/spectest/tests/runner/consensus/post_finish.go [hunk 10]
```diff
@@ -48,7 +48,7 @@ func PostFinish() *tests.MultiMsgProcessingSpecTest {
 							Root:       sha256.Sum256(testingutils.TestSyncCommitteeContributionConsensusDataByts),
 						}), nil),
 				},
-				PostDutyRunnerStateRoot: "d800b0a2047ef1caa04a7a97d803558f5f5e3c6efb5c6f0e28c94023d3676609",
+				PostDutyRunnerStateRoot: "dbc5a425e55618d83b245a71e1858419f8cd6d4b1b1022b68403da001c950601",
```
[the remark this answers is not in the document]
Comment (nkryuchkov) [reply ssv/spectest/tests/runner/consensus/post_finish.go:51]: What should I change here? 

As I understand it creates a finished (decided) runner and then processes a commit message expecting an invalid proposal error because it doesn't process any proposal messages. 

Should I add a proposal message before the commit message or delete this test or change the logic in another way?

File: qbft/spectest/tests/roundchange/f1_different_future_rounds.go (hunk 5 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/roundchange/f1_different_future_rounds.go:33]: Upon change round partial quorum we create a round change message (https://github.com/bloxapp/ssv-spec/blob/ssz-fix-remaining-tests-issues/qbft/round_change.go#L332) and call `getRoundChangeData` where `state.LastPreparedRound` is `qbft.NoRound` (because `state.LastPreparedRound` is set only in `uponPrepare` (https://github.com/bloxapp/ssv-spec/blob/ssz-fix-remaining-tests-issues/qbft/prepare.go#L35)), so it returns `[32]byte{}` as a root for the created round change message (https://github.com/bloxapp/ssv-spec/blob/ssz-fix-remaining-tests-issues/qbft/round_change.go#L301).

Should we also store all prepare messages in `InputMessages`?

File: qbft/spectest/tests/roundchange/f1_different_future_rounds_not_prepared.go (hunk 6 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/roundchange/f1_different_future_rounds_not_prepared.go:26]: same as https://github.com/bloxapp/ssv-spec/pull/162/files#r1120128955

File: ssv/spectest/tests/runner/consensus/post_finish.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (alonmuroch) [reply ssv/spectest/tests/runner/consensus/post_finish.go:51]: If that's what it's testing then it's not the purpose of the test.
The purpose is to test a consensus message post finish

File: types/testingutils/runner.go [hunk 11]
```diff
@@ -221,8 +222,8 @@ var SSVDecidingMsgs = func(consensusData *types.ConsensusData, ks *TestKeySet, r
 }
 
 var DecidingMsgsForHeight = func(consensusData *types.ConsensusData, msgIdentifier []byte, height qbft.Height, keySet *TestKeySet) []*qbft.SignedMessage {
-	r, _ := consensusData.HashTreeRoot()
 	fullData, _ := consensusData.MarshalSSZ()
+	r := sha256.Sum256(fullData)
```
Comment (alonmuroch) [inline types/testingutils/runner.go:226]: should be 
byts, _ := consensusData.Encode()
r, _ := qbft.HashDataRoot(byts)
fullData, _ := consensusData.MarshalSSZ()

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: types/testingutils/runner.go (hunk 11 above)
Comment (nkryuchkov) [reply types/testingutils/runner.go]: @alonmuroch fixed

File: ssv/spectest/tests/runner/consensus/post_finish.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (nkryuchkov) [reply ssv/spectest/tests/runner/consensus/post_finish.go]: @alonmuroch Could you please have a look at this test again? I pushed changes in https://github.com/bloxapp/ssv-spec/pull/162/commits/c6ec2c5809a06d4643c25a00e3fdf30ea8f0f259 and https://github.com/bloxapp/ssv-spec/pull/162/commits/5a5604d0ed695681d4fe79a503d4c59a39dcd05a. Does it look correct now?

[the remark this answers is not in the document]
Comment (alonmuroch) [reply ssv/spectest/tests/runner/consensus/post_finish.go]: the test tests a commit message after runner finished, we don't send a commit msg but rather a proposal so it doesn't test what the test should be testing

File: qbft/spectest/tests/messages_spectest.go (hunk 1 above)
Comment (alonmuroch) [reply qbft/spectest/tests/messages_spectest.go:28]: mm not sure what it's for...  I think we can remove it

File: qbft/spectest/tests/roundchange/justification_wrong_value.go (hunk 7 above)
Comment (alonmuroch) [reply qbft/spectest/tests/roundchange/justification_wrong_value.go:3]: we do, please reinstate the test

File: ssv/spectest/tests/runner/consensus/post_finish.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (nkryuchkov) [reply ssv/spectest/tests/runner/consensus/post_finish.go]: @alonmuroch Should the test be testing a **multi-signer** commit message after the runner has finished? With a regular commit message we'd get a `did not receive proposal for this round` error

File: qbft/spectest/tests/roundchange/justification_wrong_value.go (hunk 7 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/roundchange/justification_wrong_value.go]: Done

File: qbft/spectest/tests/messages_spectest.go (hunk 1 above)
Comment (nkryuchkov) [reply qbft/spectest/tests/messages_spectest.go]: Removed

File: ssv/spectest/tests/runner/consensus/post_finish.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (nkryuchkov) [reply ssv/spectest/tests/runner/consensus/post_finish.go]: I pushed the changes that use multi-signer commit message, could you please review?
